### PR TITLE
Deprecate TrainableDropout

### PR DESCRIPTION
**Pull request by codemzs**
This change removes TrainableDropout because Dropout(12) supports training mode.

### CODE REVIEW

**Comment by ytaous** [comment]
how about "Deprecate" instead of "Farewell" :-)

**Comment by codemzs** [comment]
@ytaous I was being funny but will change :)

**Comment by ytaous** [comment]
@manashgoswami 

---
In reply to: [448266439](https://github.com/microsoft/onnxruntime/pull/4501#issuecomment-448266439) [](ancestors = 448266439)

**Comment by ytaous** [comment]
:joy: 

---
In reply to: [659045182](https://github.com/microsoft/onnxruntime/pull/4501#issuecomment-659045182) [](ancestors = 659045182)

File: onnxruntime/core/providers/cpu/nn/dropout_op.h [hunk 1]
```diff
@@ -65,8 +65,7 @@ Status Dropout<T1, T2, trainable_dropout>::Compute(OpKernelContext* context) con
   ORT_ENFORCE(!mask || mask->Shape() == X_shape, "X and mask should have the same shape");
 
   const Tensor* training_mode = context->Input<Tensor>(2);
-  if ((0 == ratio_value /*Backward compat with TrainableDropout*/) ||
-      !trainable_dropout && (training_mode == nullptr || *(training_mode->Data<bool>()) == false)) {
+  if ((0 == ratio_value) || (training_mode == nullptr || *(training_mode->Data<bool>()) == false)) {
```
Comment (ytaous) [inline onnxruntime/core/providers/cpu/nn/dropout_op.h:68]: >0 == ratio_value) [](start = 7, length = 17)

it seems this is for TrainableDropout only, is it?

File: onnxruntime/core/providers/cuda/nn/dropout.h [hunk 2]
```diff
@@ -80,8 +77,7 @@ Status Dropout<trainable_dropout>::ComputeInternal(OpKernelContext* context) con
 
   const Tensor* training_mode = context->Input<Tensor>(2);
   //Check for inference mode.
-  if ((0 == ratio_data /*Backward compat with TrainableDropout*/) ||
-      (!trainable_dropout && (training_mode == nullptr || *(training_mode->Data<bool>()) == false))) {
+  if ((0 == ratio_data) ||(training_mode == nullptr || *(training_mode->Data<bool>()) == false)) {
```
Comment (ytaous) [inline onnxruntime/core/providers/cuda/nn/dropout.h:80]: >(0 == ratio_data) [](start = 6, length = 17)

same question

**Comment by ytaous** [comment]
@manashgoswami - do we know what models they are using? We have converter for megatron/gpt-2 and Bert. 

---
In reply to: [448266439](https://github.com/microsoft/onnxruntime/pull/4501#issuecomment-448266439) [](ancestors = 448266439)

File: onnxruntime/core/providers/cuda/nn/dropout.h (hunk 2 above)
Comment (codemzs) [reply onnxruntime/core/providers/cuda/nn/dropout.h:80]: this also applies to Dropout-12 when training_mode is true and ratio is set to zero (unlikely scenario), in which case we should short-circuit the calculation and take this branch.

File: onnxruntime/core/providers/cpu/nn/dropout_op.h (hunk 1 above)
Comment (codemzs) [reply onnxruntime/core/providers/cpu/nn/dropout_op.h:68]: this also applies to Dropout-12 when training_mode is true and ratio is set to zero (unlikely scenario), in which case we should short-circuit the calculation and take this branch.

**Comment by codemzs** [comment]
I'm seeing a failure in a non-optional CI (Win_GPU_Training RelWithDebInfo) that seems not related to my change and this seems to be coming from model zoo, moreover, I see this failure in other commits that have _already_ been merged into master which gives me confidence this is _not_ related to my change.

1: [ONNXRuntimeError] : 1 : FAIL : Non-zero status code returned while running Concat node. Name:'TFNodes/yolo_evaluation_layer_1/concat_3' Status Message: Not satisfied: dim_value == inputs_0_dims[axis_index]
1: concat.cc:79 onnxruntime::ConcatBase::PrepareForComputeNon concat axis dimensions must match: Axis 0 has mismatched dimensions of 676 and 26
1: [  FAILED  ] ModelTests/ModelTest.Run/796, where GetParam() = L"cuda_..\\models\\opset10\\yolov3\\yolov3.onnx" (2630 ms)

CC: @ytaous

**Comment by snnn** [comment]
Hi 
The test model is 20MB large,  the next please consider put such large file in another place, for example, azure blob storage.

**Comment by codemzs** [comment]
@snnn I didn't add any new model, just modified the existing one.

**Pull request merged**